File: problems/leetcode/3-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        n = len(s)
        i = ans = 0
        d = {}
        for j in range(n):
            if s[j] in d:
                i = max(d[s[j]], i)
            ans = max(ans, j - i + 1)
            d[s[j]] = j + 1
        return ans

"""
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        n = len(s)
        left = ans = 0
        found_chars = [0] * 128

        for right in range(n):
            r = s[right]
            found_chars[ord(r)] += 1
            while found_chars[ord(r)] > 1:
                l = s[left]
                found_chars[ord(l)] -= 1
                left += 1
            ans = max(ans, right - left + 1)
        return ans
"""

File: problems/leetcode/3-longest-substring-without-repeating-characters.cpp
class Solution {
   public:
    int lengthOfLongestSubstring(string s) {
        int n = s.size(), max_length = 0;
        if (n == 0) {
            return max_length;
        }
        unordered_map<char, int> indexes;
        for (int left = 0, right = 0; right < n; right++) {
            if (indexes.find(s[right]) != indexes.end()) {
                left = max(left, indexes[s[right]] + 1);
            }
            indexes[s[right]] = right;
            max_length = max(max_length, right - left + 1);
        }
        return max_length;
    }
};
